fe/fe-core/src/main/java/com/starrocks/sql/optimizer/statistics/StatisticsCalcUtils.java

```diff
@@ -267,7 +268,7 @@ private static long deltaRows(Table table, long totalRowCount) {
         Map<Long, Optional<Long>> tableStatisticMap = GlobalStateMgr.getCurrentState().getStatisticStorage()
                 .getTableStatistics(table.getId(), table.getPartitions());
 
-        for (Partition partition : table.getPartitions()) {
+        for (Partition partition : ImmutableList.copyOf(table.getPartitions())) {
```

Comment:
I'm afraid it cannot resolve the problem. Copy the list without any lock is still unsafe.

The real issue is the `table.partitions` would be modified by the other threads like `DynamicPartitionScheduler` or even `ALTER TABLE`, but sometimes we just read the `table.getPartitions` without take any lock. For most case, the `table.partitions` is protected by the lock, and it's the guideline to make it concurrent-safe.

For this particular case, i would suggest take the table before accessing `table.partitions` to avoid this CME.

Reply:
The function `table.partitions`  is to get the underlying variable:`idToPartition`, and the dropPartition also change the underlying variable:`idToPartition`,so if call `ImmutableList.copyOf` to copy  `idToPartition`  to make it immutable. so whenever dropPartition happen, the enhanced for loop to partitions never been broken,
 enhanced for loop to partitions is like blow:
```
 for (Partition partition : ImmutableList.copyOf(table.getPartitions())) {
            long partitionRowCount;
```

Reply:
> I'm afraid it cannot resolve the problem. Copy the list without any lock is still unsafe.
> 
> The real issue is the `table.partitions` would be modified by the other threads like `DynamicPartitionScheduler` or even `ALTER TABLE`, but sometimes we just read the `table.getPartitions` without take any lock. For most case, the `table.partitions` is protected by the lock, and it's the guideline to make it concurrent-safe.
> 
> For this particular case, i would suggest take the table before accessing `table.partitions` to avoid this CME.

Done, please take a review if you have time.